src/utils/DownloadManager.js

```diff
@@ -0,0 +1,241 @@
+    return (secrets && secrets.GITHUB_TOKEN) || (request.headers ? request.headers['x-github-token'] : '');
+  }
+
+  computeGithubURI(owner, repo, ref, path) {
+    const rootURI = URI.parse(this.githubRootPath);
+    const rootPath = rootURI.path;
+    // remove double slashes
+    const fullPath = `${rootPath}/${owner}/${repo}/${ref}/${path}`.replace(
+      /\/+/g,
+      '/',
+    );
+
+    return URI.serialize({
+      scheme: rootURI.scheme,
+      host: rootURI.host,
+      port: rootURI.port,
+      path: fullPath,
+    });
+  }
+
+  /**
+   * Schedules a task that fetches a web resource.
+   * @param {object} opts options.
+   * @param {object} opts.uri URI to download
+   * @param {object} opts.options Fetch options.
+   * @param {string} opts.id Some id to later identify the task.
+   * @param {number} opts.timeout Override global timeout
+   * @param {boolean} opts.errorOn404 Treat 404 as error.
+   * @return {Promise} Promise that resolves with the response.
+   */
+
+  fetch(opts) {
+    const { request = {}, logger } = this._action;
+    const context = this._context;
+
+    const { uri, options = {} } = opts;
+    if (!uri) {
+      throw new Error('Unknown uri, cannot fetch content');
+    }
+
+    if (!('timeout' in options)) {
+      options.timeout = this.timeout;
+    }
+
+    if (!('cache' in options)) {
+      options.cache = request.headers ? request.headers['Cache-Control'] : '';
+    }
+    if (this._forceNoCache) {
+      options.cache = 'no-store';
+    }
+
+    const download = async () => {
+      logger.debug(`fetching file from ${uri}`);
+      let res = {};
+      try {
+        if (this._useRequest) {
+          res = await rp({
+            uri,
+            json: false,
+            followAllRedirects: true,
+            resolveWithFullResponse: true,
+            time: true,
+            ...options,
+          });
+          res.status = res.statusCode;
+          res.ok = true;
+          return res;
+        }
+        res = await fetch(uri, options);
+        if (res.ok) {
+          return {
+            status: res.status,
+            body: await res.text(),
+            headers: res.headers,
+          };
+        }
+      } catch (e) {
+        // special handling for timeout from rp
+        if ((e.response && e.response.elapsedTime && e.response.elapsedTime > options.timeout) || (e.cause && e.cause.code && (e.cause.code === 'ESOCKETTIMEDOUT' || e.cause.code === 'ETIMEDOUT'))) {
+          e.statusCode = 502;
+        }
+
+        res = e;
+        res.status = e.statusCode;
+        res.ok = false;
+      }
+
+      if (res.status === 404) {
+        if (!opts.errorOn404) {
+          logger.warn(`Could not find file at ${uri}`);
+          res.body = 'not found';
+          return res;
+        }
+        logger.error(`Could not find file at ${uri}`);
+        setdefault(context, 'response', {}).status = 404;
+      } else if (res.status === 502 || res instanceof TimeoutError) {
```

Comment:
`fetch` never returns a `TimeoutError`, it rejects with a `TimeoutError`. The respective check should me moved to the `catch` block above.